app/assets/javascripts/omni_nav/google-custom-search.js

```diff
@@ -0,0 +1,314 @@
+      var searchBoxDiv = $searchBox[0];
+      var searchResultsDiv = $searchResults[0];
+
+      var searchBoxConfig = {
+        gname: containerId,
+        div: searchBoxDiv,
+        tag: 'searchbox',
+        attributes: {
+          enableAutoComplete: true,
+          autoCompleteMatchType: 'any',
+          resultSetSize: 6,
+          enableHistory: false
+        }
+      };
+
+      var searchResultsConfig = {
+        gname: containerId,
+        div: searchResultsDiv,
+        tag: 'searchresults',
+        attributes: {
+          linkTarget: '_self',
+          enableOrderBy: true
+        }
+      };
+
+      google.search.cse.element.render(searchBoxConfig, searchResultsConfig);
+    };
+
+    var updateSearchResultsHeight = function() {
+      $searchResultsContainer.height($(window).height());
+    };
+
+    var onSearchEnter = function(e) {
+      // The autocomplete container is not present until the user starts typing
+      // Add click event listener the first time it shows up
+      bindAutocomplete();
+
+      var resultsRequested = e.type == 'click' || e.which == ENTER_KEY;
+      var searchTermsPresent = gcsElement.getInputQuery().length > 0;
+      if( resultsRequested && searchTermsPresent ) {
+        showResults();
+      }
+    };
+
+    var onSearchFilterClick = function() {
+      $selectedSearchFilter.text($(this).text());
+    };
+
+    var onSearchResultsResize = function() {
+      clearTimeout(resizeTimeoutId);
+      resizeTimeoutId = setTimeout(updateSearchResultsHeight, 500);
+    };
+
+    var bindAutocomplete = function() {
+      // Autocomplete map. This seems very fragile as it requires Google to use
+      // the same hardcorded selector each time. But it is how this was originally coded
+      // and so far has proved reliable so I am keeping it for now.
+      var contextSelectorMap = {
+        // omniNavContext: selector
+        'primary': 'table.gstl_50.gssb_c',
+        'utility': 'table.gstl_51.gssb_c'
+      };
+
+      // Bind only once: use a data flag to track binding.
+      var autoCompleteIsBound = $container.data('autoCompleteIsBound');
+      if ( autoCompleteIsBound ) return;
+
+      // This is required for selecting the Autocomplete block associated with
+      // SearchComponent. Since these are added post-loading by Google and are attached
+      // as absolute elements directly to the body, the parent cannot be deduced
+      // from the object itself. Checked Google API. No accessor. This is unfortunate.
+      var omniNavContext = ( containerId.indexOf("utility") >= 0 ) ? 'utility' : 'primary';
+      var autoCompleteSelector = contextSelectorMap[omniNavContext];
+      $container.data('autoCompleteIsBound', true);
+
+      // Bind callback: when autocomplete option is clicked, show results. Interval
+      // allows time for autoCompleteTable to be added to DOM.
+      var intervalId = setInterval(function() {
```

Comment:
This interval checking wasn't present in the original code but in testing on the local sample page the `$autoCompleteTable` would sometimes not be found. It's ugly but stopped the problem.

Have anyone noticed or received any reports of search results sometimes failing to load?

Reply:
@tatwell In regards to this comment and the one in the code, `Interval allows time for autoCompleteTable to be added to DOM`, that's what `setTimeout` was for here:
```
var onAutoCompleteClick = function() {
    setTimeout(showResults, 100);
};
```
Was that not working anymore? I never had an issue with it failing to load, while the timeout was there. I did test it briefly in the beginning just to be sure that it failed if the timeout was taken out.

Reply:
Yes, I think there are 2 potential race conditions here:

1. Bind `onAutoCompleteClick` callback to `$autoCompleteTable`: need to wait for Google to insert `$autoCompleteTable` into DOM.
1. When clicking `$autoCompleteTable`, waiting for something else to happen (Google loading results?) before showing results.

\#2 was already in the code. I tried removing both independently and got intermittent issues where an expected block would be blank or missing. So I think both are required. Using timeout is still prone to problems if the dependent event takes longer than the timeout delay.

As Mirabel noted, this is due to limitations in the Google API. It would be nice if they offered callbacks or events that could be used instead of this guesswork. I suppose it reflects the fact that they don't really want you extending their UI.